rename `widgets` table to `new_widgets`

kuldeepaggarwal opened the pull request:
`widgets` table is being created in [primary_keys_test.rb](https://github.com/rails/rails/blob/master/activerecord/test/cases/primary_keys_test.rb#L256-L258) for PostgreSQLAdapter, MysqlAdapter, Mysql2Adapter only and it makes test to fail for these adapters irrespective of how many times I execute the test suite.

Before:
  `bundle exec rake mysql2:test`

```
Finished in 127.287669s, 35.5258 runs/s, 97.8885 assertions/s.

  1) Error:
PersistenceTest::SaveTest#test_save_touch_false:
ActiveModel::UnknownAttributeError: unknown attribute 'name' for #<Class:0x0000000a7d6ef0>.
    /home/kd/projects/kd-rails/activerecord/lib/active_record/attribute_assignment.rb:36:in `rescue in _assign_attribute'
    /home/kd/projects/kd-rails/activerecord/lib/active_record/attribute_assignment.rb:34:in `_assign_attribute'
    /home/kd/projects/kd-rails/activemodel/lib/active_model/attribute_assignment.rb:40:in `block in _assign_attributes'
    /home/kd/projects/kd-rails/activemodel/lib/active_model/attribute_assignment.rb:39:in `each'
    /home/kd/projects/kd-rails/activemodel/lib/active_model/attribute_assignment.rb:39:in `_assign_attributes'
    /home/kd/projects/kd-rails/activerecord/lib/active_record/attribute_assignment.rb:26:in `_assign_attributes'
    /home/kd/projects/kd-rails/activemodel/lib/active_model/attribute_assignment.rb:33:in `assign_attributes'
    /home/kd/projects/kd-rails/activerecord/lib/active_record/core.rb:293:in `initialize'
    /home/kd/projects/kd-rails/activerecord/lib/active_record/inheritance.rb:61:in `new'
    /home/kd/projects/kd-rails/activerecord/lib/active_record/inheritance.rb:61:in `new'
    /home/kd/projects/kd-rails/activerecord/lib/active_record/persistence.rb:50:in `create!'
    /home/kd/projects/kd-rails/activerecord/test/cases/persistence_test.rb:913:in `test_save_touch_false'

4522 runs, 12460 assertions, 0 failures, 1 errors, 4 skips
```

After:
  `bundle exec rake mysql2:test`

```
   Finished in 135.785086s, 33.3026 runs/s, 91.7774 assertions/s.

   4522 runs, 12462 assertions, 0 failures, 0 errors, 4 skips
```

P.S. I don't know how Travis is so lucky every time.

senny commented:
@kuldeepaggarwal This patch does not feel right. If a test-case creates a table it's responsible to cleanup afterwards. The linked `primary_key_test.rb` does that in the `teardown` block: https://github.com/rails/rails/blob/master/activerecord/test/cases/primary_keys_test.rb#L263

Instead of simply changing the name to avoid the collision we should investigate where the leak is happening.

kuldeepaggarwal commented:
@senny: so is there any settings in the test suite which runs test cases in parallel as it might be some issue of parallelism!!

senny commented:
@kuldeepaggarwal I don't think so. I imagine there would be a bunch of failing tests if we ran the AR suite in parallel.

kuldeepaggarwal commented:
@senny: Test case also passes if I use `widget.reset_column_information`.

rafaelfranca commented:
@kuldeepaggarwal could you take another look. I agree with @senny this patch should not be required.

kuldeepaggarwal commented:
@rafaelfranca: will give another shot, but I am surprise why tests are passing if I use `widget.reset_column_information` in this test itself without renaming the table name.

senny commented on activerecord/test/cases/persistence_test.rb:
```diff
@@ -924,7 +924,7 @@ def test_save_touch_false
       assert_equal instance.updated_at, updated_at
     ensure
-      ActiveRecord::Base.connection.drop_table :widgets
+      ActiveRecord::Base.connection.drop_table widget.table_name
```
Instead of changing the table name to prevent the collision, we should use `widget.reset_column_information`. The thing is that this method clears the table cache, which is connection wide.

kuldeepaggarwal replied:
:+1:

kuldeepaggarwal commented:
Done, updated the PR.

senny commented:
@kuldeepaggarwal the `reset_column_information` should happen in an `ensure` as cleanup, not in advance:

* here: https://github.com/rails/rails/blob/e2a96c071216d14659294a5a7645878656c30916/activerecord/test/cases/primary_keys_test.rb#L263
* here: https://github.com/rails/rails/blob/e2a96c071216d14659294a5a7645878656c30916/activerecord/test/cases/persistence_test.rb#L927

Cleaning up before the test misses the opportunity to prevent the state leaks in the first place.

kuldeepaggarwal commented:
Good point :+1: . Thanks for pointing it out.

kuldeepaggarwal commented:
done. :+1:

senny commented on activerecord/test/cases/persistence_test.rb:
```diff
@@ -910,6 +910,7 @@ def test_save_touch_false
         self.table_name = :widgets
       end
 
+
```
:scissors:

kuldeepaggarwal replied:
done.

kuldeepaggarwal committed:
use Model.reset_column_information to clear table cache connection wide.

    `widgets` table is being created in `primary_keys_test.rb` for PostgreSQLAdapter, MysqlAdapter, Mysql2Adapter
    and it makes test to fail earlier.

    Before:
      `bundle exec rake mysql2:test`

    ```
    Finished in 127.287669s, 35.5258 runs/s, 97.8885 assertions/s.

      1) Error:
    PersistenceTest::SaveTest#test_save_touch_false:
    ActiveModel::UnknownAttributeError: unknown attribute 'name' for #<Class:0x0000000a7d6ef0>.
        /home/kd/projects/kd-rails/activerecord/lib/active_record/attribute_assignment.rb:36:in `rescue in _assign_attribute'
        /home/kd/projects/kd-rails/activerecord/lib/active_record/attribute_assignment.rb:34:in `_assign_attribute'
        /home/kd/projects/kd-rails/activemodel/lib/active_model/attribute_assignment.rb:40:in `block in _assign_attributes'
        /home/kd/projects/kd-rails/activemodel/lib/active_model/attribute_assignment.rb:39:in `each'
        /home/kd/projects/kd-rails/activemodel/lib/active_model/attribute_assignment.rb:39:in `_assign_attributes'
        /home/kd/projects/kd-rails/activerecord/lib/active_record/attribute_assignment.rb:26:in `_assign_attributes'
        /home/kd/projects/kd-rails/activemodel/lib/active_model/attribute_assignment.rb:33:in `assign_attributes'
        /home/kd/projects/kd-rails/activerecord/lib/active_record/core.rb:293:in `initialize'
        /home/kd/projects/kd-rails/activerecord/lib/active_record/inheritance.rb:61:in `new'
        /home/kd/projects/kd-rails/activerecord/lib/active_record/inheritance.rb:61:in `new'
        /home/kd/projects/kd-rails/activerecord/lib/active_record/persistence.rb:50:in `create!'
        /home/kd/projects/kd-rails/activerecord/test/cases/persistence_test.rb:913:in `test_save_touch_false'

    4522 runs, 12460 assertions, 0 failures, 1 errors, 4 skips
    ```

    After:
      `bundle exec rake mysql2:test`

    ```
       Finished in 135.785086s, 33.3026 runs/s, 91.7774 assertions/s.

       4522 runs, 12462 assertions, 0 failures, 0 errors, 4 skips
    ```

senny commented:
@kuldeepaggarwal thank you :yellow_heart:

kuldeepaggarwal commented:
@senny Thank you..
